### PR TITLE
Enh/file name validation

### PR DESCRIPTION
**Pull request by Cena-Studio**
# Overview

Based on the current `silo file` syntax, the colon in the filename makes it difficult to distinguish whether the content before the colon is a silo name or part of the file name. To solve this problem, colon and space are no longer allowed to be included in a filename, which will be validated during `silo file push` operation.

# Major Changes

- Modified the algorithm in `FilePush#run` to correctly parse the silo name and destination argument and validate the destination
- A corresponding `InvalidFileNameError` is defined in errors.rb

### CODE REVIEW

File: lib/silo/commands/file_push.rb [hunk 1]
```diff
@@ -40,14 +40,12 @@ def run
...
-        if args[1]&.match(/^[^:]*:[^:]*$/)
-          silo_name, dest = args[1].split(":").map(&:to_s)
-        else
-          silo_name = Silo.default
-          dest = args[1] || ''
-        end
-
-        dest ||= ''
+        # standardize args[1] to [silo_name, dest]
+        split_args = args[1]&.split(":", 2).map(&:to_s) || ['']
```
Comment (VoxSecundus) [inline lib/silo/commands/file_push.rb:44]: This doesn't work. If `args[1]` is `nil`, you will get an error upon attempting to call `map` on it (`split` will return `nil` because of the nilsafe operator).

Also, the following test looks incorrect:

```ruby
2.7.1 :017 > def test(str)
2.7.1 :018 >   x = str&.split(':', 2).map(&:to_s) || ['']
2.7.1 :019 >   x.tap { |f| f.unshift('jack-silo') if x.size == 1 }
2.7.1 :020 > end
 => :test
2.7.1 :024 > test(':f')
 => ["", "f"]
```

`:f` is an invalid input, as it contains no silo name and therefore a filename with a colon in it`.

File: lib/silo/errors.rb [hunk 2]
```diff
@@ -32,4 +32,5 @@ module FlightSilo
   NoSuchSiloError = Class.new(RuntimeError)
   NoSuchDirectoryError = Class.new(RuntimeError)
   NoSuchFileError = Class.new(RuntimeError)
+  InvalidFileNameError = Class.new(RuntimeError)
```
Comment (VoxSecundus) [inline lib/silo/errors.rb:35]: Well done in thinking to add a new error class for this.

**Review by VoxSecundus — CHANGES_REQUESTED** [review]

File: lib/silo/commands/file_push.rb [hunk 3]
```diff
@@ -40,14 +40,13 @@ def run
...
 
-        if args[1]&.match(/^[^:]*:[^:]*$/)
-          silo_name, dest = args[1].split(":").map(&:to_s)
-        else
-          silo_name = Silo.default
-          dest = args[1] || ''
-        end
-
-        dest ||= ''
+        raise InvalidFileNameError, "Target destination '#{args[1]}' contains invalid symbol" unless args[1].nil? || args[1].match?(/^[^:]+(:[^:]+)?$/)
```
Comment (VoxSecundus) [inline lib/silo/commands/file_push.rb:43]: This still allows the pushing of files with colons in their filenames, etc:
```
[flight@jack-tests-profile (my-cluster) [login1] flight-silo]$ flexec ruby bin/silo set-default jack-silo
Default silo set to: jack-silo
[flight@jack-tests-profile (my-cluster) [login1] flight-silo]$ flexec ruby bin/silo file push :test
Local file '/home/flight/flight-silo/:test' copied to remote '/:test'
```

Trying to pull this file later raises an error:
```
[flight@jack-tests-profile (my-cluster) [login1] flight-silo]$ flexec ruby bin/silo file pull :test
silo: Silo '' not found
```

**Comment by VoxSecundus** [comment]
This is _almost_ there; there's still a case where, if I run `file push README.md jack-test:`, an error is raised because the second argument has an illegal character (`:`), but that's how we specify a repo to push to. This is a feature with a lot of edge cases, and I'll come back to it momentarily to look for a solution.

File: lib/silo/commands/file_push.rb [hunk 4]
```diff
@@ -94,6 +96,19 @@ def run
...
+        Dir.foreach(directory) do |item|
+          next if item == '.' || item == '..'
+          item_path = File.join(directory, item)
+          raise InvalidFileNameError, "Source file or directory '#{item_path}' contains invalid symbol" if item.match?(/[: ]/)
+          
+          if File.directory?(item_path)
+            traverse_validation(item_path)
+          end
+        end
+      end
```
Comment (VoxSecundus) [inline lib/silo/commands/file_push.rb:111]: Good idea here.

**Review by VoxSecundus — APPROVED** [review]
This works great now. I've pushed a commit to revert the behaviour in how we prepare Silo types, because it had broken. :shipit:
